### PR TITLE
masync: make future safe to poll when complete

### PR DESCRIPTION
**Pull request by pbalcer**
This patch makes it safe to poll on completed futures,
simplifying group handling of futures (e.g., in runtimes).
I've also added a new basic test and a few tiny helper APIs.

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/miniasync/61)

### CODE REVIEW

File: tests/future/test_future.cmake [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+# an example for the basic test case
+
+include(${SRC_DIR}/cmake/test_helpers.cmake)
+
+setup()
+
+# The expected input can be provided to the execute function,
+# which will check if the return code of the binary file matches.
+execute(0 ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${BUILD}/dummy)
```
Comment (DamianDuy) [inline tests/future/test_future.cmake:12]: I think you are accidentally testing a wrong binary.

Comment (pbalcer) [reply tests/future/test_future.cmake]: So that's why it worked on the first try :D Good catch.
Fixed.

**Comment by kswiecicki** [comment]
:lgtm:

**Review by DamianDuy — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/61#-:-MxhX2S9-wnrhCnf4Y05:bnfp4nl)* status: 0 of 4 files reviewed, 1 unresolved discussion (waiting on @DamianDuy)

File: tests/future/test_future.cmake [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright 2021, Intel Corporation
```
Comment (wlemkows) [inline tests/future/test_future.cmake:2]: 22, why id does not fail?

Comment (pbalcer) [reply tests/future/test_future.cmake]: Done.

**Pull request merged**